lib/link_state.c

```diff
@@ -0,0 +1,911 @@
+		STREAM_GETL(s, ls_pref->metric);
+	if (CHECK_FLAG(ls_pref->flags, LS_PREF_SR)) {
+		STREAM_GETL(s, ls_pref->sr.sid);
+		STREAM_GETC(s, ls_pref->sr.sid_flag);
+		STREAM_GETC(s, ls_pref->sr.algo);
+	}
+
+	return ls_pref;
+
+stream_failure:
+	zlog_err("LS(%s): Could not parse Link State Prefix. Abort!", __func__);
+	XFREE(MTYPE_LS_DB, ls_pref);
+	return NULL;
+}
+
+struct ls_message *ls_parse_msg(struct stream *s)
+{
+	struct ls_message *msg;
+
+	msg = XCALLOC(MTYPE_LS_DB, sizeof(struct ls_message));
+	if (msg == NULL)
+		return NULL;
+
+	/* Read LS Message header */
+	STREAM_GETC(s, msg->event);
+	STREAM_GETC(s, msg->type);
+	STREAM_GET(&msg->remote_id, s, sizeof(struct ls_node_id));
+
+	/* Read Message Payload */
+	switch (msg->type) {
+	case LS_MSG_TYPE_NODE:
+		msg->data.node = ls_parse_node(s);
+		break;
+	case LS_MSG_TYPE_ATTRIBUTES:
+		msg->data.attr = ls_parse_attributes(s);
+		break;
+	case LS_MSG_TYPE_PREFIX:
+		msg->data.prefix = ls_parse_prefix(s);
+		break;
+	default:
+		zlog_err("Unsupported Payload");
+		goto stream_failure;
+	}
+
+	if (msg->data.node == NULL || msg->data.attr == NULL
+	    || msg->data.prefix == NULL)
+		goto stream_failure;
+
+	return msg;
+
+stream_failure:
+	zlog_err("LS(%s): Could not parse LS message. Abort!", __func__);
+	XFREE(MTYPE_LS_DB, msg);
+	return NULL;
+}
+
+static int ls_format_node(struct stream *s, struct ls_node *node)
+{
+	size_t len;
+
+	/* Push Advertise node information first */
+	stream_put(s, &node->adv, sizeof(struct ls_node_id));
+
+	/* Push Flags & Origin then Node information if there are present */
+	stream_putw(s, node->flags);
+	if (CHECK_FLAG(node->flags, LS_NODE_NAME)) {
+		len = strlen(node->name);
+		stream_putc(s, len + 1);
+		stream_put(s, node->name, len);
+		stream_putc(s, '\0');
+	}
+	if (CHECK_FLAG(node->flags, LS_NODE_ROUTER_ID))
+		stream_put_ipv4(s, node->router_id.s_addr);
+	if (CHECK_FLAG(node->flags, LS_NODE_ROUTER_ID6))
+		stream_put(s, &node->router6_id, IPV6_MAX_BYTELEN);
+	if (CHECK_FLAG(node->flags, LS_NODE_FLAG))
+		stream_putc(s, node->node_flag);
+	if (CHECK_FLAG(node->flags, LS_NODE_TYPE))
+		stream_putc(s, node->type);
+	if (CHECK_FLAG(node->flags, LS_NODE_AS_NUMBER))
+		stream_putl(s, node->as_number);
+	if (CHECK_FLAG(node->flags, LS_NODE_SR)) {
+		stream_putl(s, node->srgb.lower_bound);
+		stream_putl(s, node->srgb.range_size);
+		stream_putc(s, node->srgb.flag);
+		stream_put(s, node->algo, 2);
+	}
+	if (CHECK_FLAG(node->flags, LS_NODE_SRLB)) {
+		stream_putl(s, node->srlb.lower_bound);
+		stream_putl(s, node->srlb.range_size);
+	}
+	if (CHECK_FLAG(node->flags, LS_NODE_MSD))
+		stream_putc(s, node->msd);
+
+	return 0;
+}
+
+static int ls_format_attributes(struct stream *s, struct ls_attributes *attr)
+{
+	size_t len;
+
+	/* Push Advertise node information first */
+	stream_put(s, &attr->adv, sizeof(struct ls_node_id));
+
+	/* Push Flags & Origin then LS attributes if there are present */
+	stream_putc(s, attr->flags);
```

Comment:
Should be `stream_putl(s, attr->flags)` ?

Reply:
Done